### ISSUE TITLE
runtime/cgo: `cgo.Handle` examples should highlight the requirement to manually pin the handle when the handle is used after the call ends

### ISSUE THREAD

**Issue by JAicewizard** [golang/go#68044]
The documentation on for `cgo.Handle` passes a go pointer (pointer to go memory) to C, which is only allowed when the value is pinned (unable to be freed/moved). This is the case for the duration of the call when calling C functions, however in more complex systems this is not always the case. The example should be updated, or a note added, to reflect this. Users may freely copy and paste this example not realizing this requires more attention.

As an example of this happening, this happened twice during the development of go-duckdb, once by me and once by @taniabogatsch

See example:
```
package main

/*
extern void MyGoPrint(void *context);
static inline void myprint(void *context) {
    MyGoPrint(context);
}
*/
import "C"
import (
	"runtime/cgo"
	"unsafe"
)

//export MyGoPrint
func MyGoPrint(context unsafe.Pointer) {
	h := *(*cgo.Handle)(context)
	val := h.Value().(string)
	println(val)
	h.Delete()
}

func main() {
	val := "hello Go"
	h := cgo.NewHandle(val)
	C.myprint(unsafe.Pointer(&h))
	// Output: hello Go
}
```

As the required work is small I can implement this if needed.

**Comment by mknyszek** [comment]
I see what you're saying about the example. I agree it could be clearer about the fact that `unsafe.Pointer(&h)` is not passing the handle, but a pointer to a `cgo.Handle`-sized allocation, which is not kept live in the same way as the memory referenced by the `cgo.Handle`.

I don't think the example should explicitly pin the allocation (since that's introducing a new concept in the documentation for `cgo.Handle`) but perhaps it should have a short comment explaining that `unsafe.Pointer(&h)` _might_ have to be pinned to at least call attention to it.

**Comment by JAicewizard** [comment]
Yup, that would completely solve the issue. Changing the example might be too much but this sounds perfect!

**Comment by elichai** [comment]
Isn't this a sign that maybe the design of `cgo.handle` is not great?
It would be far better if you could pass a `cgo.Handle` directly as a `*C.void` pointer, as this is very commonly used in callback based APIs where the caller calls a go callback with that pointer *after* the c call has returned

**Comment by JAicewizard** [comment]
In my usecase I made a simple generic wrapper to store values used in callbacks, (no time RN, but you can look it up in go-duckdb). I think it does its job pretty well, wouldn't mind something like it landing in `cgo` package

**Comment by ianlancetaylor** [comment]
The current `cgo.Handle` is an integer, and the Go garbage collector doesn't permit confusing integers and pointers. So I don't think we can permit the current API to seamlessly convert to a C pointer. That would make it to easy to accidentally turn it into a Go pointer.

**Comment by JAicewizard** [comment]
No I understand, I don't think "magically" changing the handle type would be good at all, but adding a new type like `PinnedHandle` could be beneficial
